components/inputs/input-date.js

```diff
@@ -20,10 +21,14 @@ export function formatISODateInUserCalDescriptor(val) {
  * A component that consists of a text input field for typing a date and an attached calendar (d2l-calendar) dropdown. It displays the "value" if one is specified, or a placeholder if not, and reflects the selected value when one is selected in the calendar or entered in the text input.
  * @fires change - Dispatched when a date is selected or typed. "value" reflects the selected value and is in ISO 8601 calendar date format ("YYYY-MM-DD").
  */
-class InputDate extends LocalizeCoreElement(LitElement) {
+class InputDate extends FormElementMixin(LocalizeCoreElement(LitElement)) {
 
 	static get properties() {
 		return {
+			/**
+			 * Indicates that the input value is invalid
+			 */
+			ariaInvalid: { type: String, attribute: 'aria-invalid', reflect: true },
```

Comment:
I know this is something that the `FormElementMixin` is setting... but seeing how it ends up working here, I'm wondering if it could instead be a non-reflected property called something like `invalid`? That way each component can do with that what they like... in input-date's case it would transfer that as `aria-invalid='true'` onto its text input. Other components might do other things.

Reply:
Its definitely open to change. This was based off of `d2l-input-text` but if thats something we're not happy with now is the time to change it.

Reply:
I think it's worth changing... otherwise we'll have custom elements sprouting `aria-invalid` attributes which don't actually mean anything. `aria-invalid` should really only apply to the underlying element(s) that are invalid -- in `d2l-text-input`'s case, that's the `<input type="text">`.

Reply:
Something that does concern me a bit is every time we make changes like these we diverge from how native and custom form elements are handled. The major cons I see with this are:
1. It increases the learning curve since.
1. It complicates `d2l-form` because there needs to be a bunch of branching logic for if an element is custom or native.
1. It will make it more difficult to ever roll out form associated custom elements if we reach a point when all browsers support them.

I think its reasonable to do in some cases but we should think really carefully when we do to make sure we're actually making the right long term decision.

Reply:
Hmm ok I wasn't aware that `aria-invalid` was a form-associated-custom-elements thing. Typically ARIA is only ever used as a crutch until standards can catch up -- they rarely actually make use of it themselves.

Reply:
Form associated custom elements basically involve extending the HTML 5 validation API to custom components so attributes that work on native components will work on custom form elements. This also includes `aria-invalid`.

It doesn't need to be a form associated custom element thing but it would likely be what we use because `:invalid` has major short-comings that make it unusable:
1. `:invalid` being added on page load for required elements.
1. Not being able to add `:invalid` for `d2l-validator-custom`'s.

I do think we can remove it, I just want to make sure we thinking about the broad picture when we make the decision.

Reply:
Right. So FACEs (haha) could essentially say... "Ignore the text input(s) inside my shadow root, they're not invalid but I am". And that might end up being accomplished using `aria-invalid`... although ideally the AOM will expose a way to say an element is invalid without using ARIA -- much the same way it's supposed to allow elements to take on particular roles without using the `role` attribute (part of ARIA).

I still feel like we're OK to rename this without deviating too much from FACE. Until FACE + the AOM exist, we'll likely need to be marshalling `aria-invalid` attributes onto nested native elements anyway.

Reply:
So just so I'm clear before I make changes - is the thought that `FormElementMixin` should have a property `invalid` that is set instead of the `this.setAttribute('aria-invalid', 'true|false');` and then we have components like `input-date` listen for that and do what they need to?

Reply:
Alternatively if we're going the route of not using `aria-invalid` it might make the most sense to combine `showValidationTooltip` / `hideValidationTooltip` and just use a single `validationError` property that can be the current error message or `null`. Then the component can choose what to do with it.

Reply:
Cool I like that 👍